main.go

```diff
@@ -279,74 +344,64 @@ func read(ctx context.Context, endpoint *url.URL, labels []prompb.Label, latency
 	return nil
 }
 
-func generate(labels []prompb.Label) *prompb.WriteRequest {
-	timestamp := time.Now().UnixNano() / int64(time.Millisecond)
-
-	return &prompb.WriteRequest{
-		Timeseries: []prompb.TimeSeries{
-			{
-				Labels: labels,
-				Samples: []prompb.Sample{
-					{
-						Value:     float64(timestamp),
-						Timestamp: timestamp,
-					},
-				},
-			},
-		},
-	}
-}
-
-func write(ctx context.Context, endpoint *url.URL, token string, wreq *prompb.WriteRequest) error {
+func write(ctx context.Context, endpoint fmt.Stringer, token string, wreq proto.Message, m metrics) error {
 	var (
 		buf []byte
 		err error
 		req *http.Request
 		res *http.Response
 	)
+
 	defer func() {
 		if err != nil {
-			remoteWriteRequests.WithLabelValues("error").Inc()
+			m.remoteWriteRequests.WithLabelValues("error").Inc()
 			return
 		}
-		remoteWriteRequests.WithLabelValues("success").Inc()
+		m.remoteWriteRequests.WithLabelValues("success").Inc()
 	}()
+
 	buf, err = proto.Marshal(wreq)
 	if err != nil {
 		return errors.Wrap(err, "marshalling proto")
 	}
+
 	req, err = http.NewRequest("POST", endpoint.String(), bytes.NewBuffer(snappy.Encode(nil, buf)))
 	if err != nil {
 		return errors.Wrap(err, "creating request")
 	}
+
 	req.Header.Add("Authorization", "Bearer "+token)
+
 	res, err = http.DefaultClient.Do(req.WithContext(ctx))
 	if err != nil {
 		return errors.Wrap(err, "making request")
 	}
+	defer res.Body.Close()
```

Comment:
Nice, but we also need to do something like `ExhaustCloseWithLogOnErr` to read all from body first. Otherwise, it will not reuse connections.